teps/draft-20200715-access-control-via-sa.md

```diff
@@ -0,0 +1,249 @@
+serve as the contact point for intercepting any create/update requests for TaskRun and PipelineRun and 
+then see if the ServiceAccount specified (or "default" if non specified) has permission to get/read any
+referenced ClusterTask/Task/Pipeline objects.
+
+A Kubernetes RBAC Subject Access Review (SAR) check will be executed to see if the ServiceAccount in question
+is part of a Role/RoleBinding or ClusterRole/ClusterRoleBinding that allows access to the referenced 
+ClusterTask, Task, or PipelineTask.  If access is denied, the admission webhook will prevent creation/update
+from completing per the contract specified by the Kubernetes Validating Admission Webhook API.
+
+By default, this new webhook will not perform any checks.  However, on each request it will interrogate 
+a documented environment variable (which is specified on the Tekton Webhook Deployment).  If with this environment
+variable:
+
+- "all" is specified, preceded/followed by a space, factoring in beginning and end of the value of the environment 
+variable
+- or a space delimited list of names corresponding to namespaces is provide (where "all" could be one of those 
+entries as well, hence turning the feature on for every namespace)
+
+then the new webhook will start enforcement on the incoming PipelineRun or TaskRun.
+
+A second environment variable around enforcement of cluster scoped resources, namespace scoped resources, or both,
+could also be employed if that level of configuration enablement is desired.   
+
+### User Stories (optional)
+
+As a Kubernetes cluster administrator, I want to dynamically modify Tekton's configuration such that enforcement
+of required permissions to get/read cluster scoped resources like ClusterTask from a PipelineRun or TaskRun is 
+possible, using the ServiceAccount (specified or default) as the actor in the permission check, and where 
+enablement can be done on a global basis or per namespace basis, and where enablement can vary for cluster scoped
+resources vs. namespace scoped resources.
+
+As a Kubernetes cluster administrator, I want to dynamically modify Tekton's configuration such that enforcement
+of required permissions to get/read namespace scoped resources like Task from a TaskRun or Task/Pipeline from a 
+PipelineRun is possible, where all objects are in the same namespace, using the ServiceAccount (specified or 
+default) as the actor in the permission check, and where enablement can be done on a global basis or per 
+namespace basis, and where enablement can vary for cluster scoped resources vs. namespace scoped resources. 
+
+### Notes/Constraints/Caveats (optional)
+
+
+### Risks and Mitigations
+
+"Security first" is a mantra often heard amongst Kubernetes cluster administrators.  The fact that some form of 
+security enforcement is not on by default very possibly will come under scrutiny.  This proposal for now does not go that 
+far, in the interest of minimizing disruption of the current adoption of Tekton.
+
```

Comment:
Kubernetes in general has suffered from a lot of security issues caused by trying to add features after the GA release. When these security improvements would cause breakages to existing use-cases, they have remained off by default.

I understand this approach, but given we're still pre-GA it might be worth being a bit more aggressive in rolling out some of these features. WDYT?

We would still need to evaluate each change on a case-by-case basis, but I don't think we need to necessarily always prioritize minimizing disruption.

Reply:
I agree, pre-GA is definitely the time to make disruptive changes!

> In reality I assumed resistance to that so proposed something less potentially controversial.  I do think allowing
to turn off on a per namespace basis could assist in migrating to the new norms.

When I push back against proposals, I am motivated by a desire to make sure we fully explore the problem(s) we're trying to solve and be very purposeful about what we implement.

I don't mind disrupting the norms we've established so far, but I want to make sure that we want to before we do it.